File: sdks/go/dcp/v2/dual_hash.go
package v2

import (
	"crypto/sha256"
	"encoding/hex"

	"golang.org/x/crypto/sha3"
)

// DualHash holds both SHA-256 and SHA3-256 hex digests.
type DualHash struct {
	SHA256  string `json:"sha256"`
	SHA3256 string `json:"sha3_256"`
}

// SHA256Hex returns the hex-encoded SHA-256 digest of data.
func SHA256Hex(data []byte) string {
	h := sha256.Sum256(data)
	return hex.EncodeToString(h[:])
}

// SHA3256Hex returns the hex-encoded SHA3-256 digest of data.
func SHA3256Hex(data []byte) string {
	h := sha3.Sum256(data)
	return hex.EncodeToString(h[:])
}

// ComputeDualHash returns SHA-256 and SHA3-256 digests for raw bytes.
func ComputeDualHash(data []byte) DualHash {
	return DualHash{
		SHA256:  SHA256Hex(data),
		SHA3256: SHA3256Hex(data),
	}
}

// DualHashCanonical computes the dual hash of a canonical JSON string.
func DualHashCanonical(canonicalJSON string) DualHash {
	return ComputeDualHash([]byte(canonicalJSON))
}

// merkleRootHex reduces a layer of hex-encoded leaf hashes to a single root
// using the provided hashFn. Odd layers are balanced by duplicating the last
// leaf, matching the Python/TS/WASM behaviour.
func merkleRootHex(leaves []string, hashFn func([]byte) string) (string, error) {
	if len(leaves) == 0 {
		return "", nil
	}
	layer := make([][]byte, len(leaves))
	for i, s := range leaves {
		b, err := hex.DecodeString(s)
		if err != nil {
			return "", err
		}
		layer[i] = b
	}
	for len(layer) > 1 {
		if len(layer)%2 == 1 {
			layer = append(layer, layer[len(layer)-1])
		}
		next := make([][]byte, 0, len(layer)/2)
		for i := 0; i < len(layer); i += 2 {
			combined := append([]byte{}, layer[i]...)
			combined = append(combined, layer[i+1]...)
			h := hashFn(combined)
			decoded, err := hex.DecodeString(h)
			if err != nil {
				return "", err
			}
			next = append(next, decoded)
		}
		layer = next
	}
	return hex.EncodeToString(layer[0]), nil
}

// DualMerkleRoot reduces two parallel layers (SHA-256 + SHA3-256) of leaf
// commitments into a single pair of Merkle roots. Returns a zero-value
// DualHash and ok=false when the leaf list is empty. Byte-identical to
// dualMerkleRoot (TypeScript) and dual_merkle_root (Python).
func DualMerkleRoot(leaves []DualHash) (DualHash, bool) {
	if len(leaves) == 0 {
		return DualHash{}, false
	}
	sha256Leaves := make([]string, len(leaves))
	sha3Leaves := make([]string, len(leaves))
	for i, l := range leaves {
		sha256Leaves[i] = l.SHA256
		sha3Leaves[i] = l.SHA3256
	}
	sha256Root, err := merkleRootHex(sha256Leaves, SHA256Hex)
	if err != nil {
		return DualHash{}, false
	}
	sha3Root, err := merkleRootHex(sha3Leaves, SHA3256Hex)
	if err != nil {
		return DualHash{}, false
	}
	return DualHash{SHA256: sha256Root, SHA3256: sha3Root}, true
}

File: sdks/rust/src/v2/dual_hash.rs
use serde::{Deserialize, Serialize};
use sha2::{Sha256, Digest};
use sha3::Sha3_256;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DualHash {
    pub sha256: String,
    pub sha3_256: String,
}

pub fn sha256_hex(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    hex::encode(hasher.finalize())
}

pub fn sha3_256_hex(data: &[u8]) -> String {
    let mut hasher = Sha3_256::new();
    hasher.update(data);
    hex::encode(hasher.finalize())
}

pub fn dual_hash(data: &[u8]) -> DualHash {
    DualHash {
        sha256: sha256_hex(data),
        sha3_256: sha3_256_hex(data),
    }
}

pub fn dual_hash_canonical(canonical_json: &str) -> DualHash {
    dual_hash(canonical_json.as_bytes())
}

/// Reduce a layer of hex-encoded leaf hashes to a single Merkle root using
/// the given hasher. Odd layers are balanced by duplicating the last leaf,
/// matching the Python/TS/WASM behaviour.
fn merkle_root_hex<D: Digest>(leaves: &[String]) -> Option<String> {
    if leaves.is_empty() {
        return None;
    }
    let mut layer: Vec<Vec<u8>> = leaves
        .iter()
        .map(|s| hex::decode(s).unwrap_or_default())
        .collect();
    while layer.len() > 1 {
        if layer.len() % 2 == 1 {
            layer.push(layer.last().cloned().unwrap_or_default());
        }
        let mut next: Vec<Vec<u8>> = Vec::with_capacity(layer.len() / 2);
        for pair in layer.chunks(2) {
            let mut hasher = D::new();
            hasher.update(&pair[0]);
            hasher.update(&pair[1]);
            next.push(hasher.finalize().to_vec());
        }
        layer = next;
    }
    Some(hex::encode(&layer[0]))
}

/// Compute dual Merkle roots from a list of dual-hash leaves.
///
/// Each leaf carries both a SHA-256 and a SHA3-256 commitment; this
/// function reduces the two layers in parallel and returns the pair of
/// roots. Returns `None` when the leaf list is empty. Byte-identical to
/// `dualMerkleRoot` (TypeScript) and `dual_merkle_root` (Python).
pub fn dual_merkle_root(leaves: &[DualHash]) -> Option<DualHash> {
    if leaves.is_empty() {
        return None;
    }
    let sha256_leaves: Vec<String> = leaves.iter().map(|l| l.sha256.clone()).collect();
    let sha3_leaves: Vec<String> = leaves.iter().map(|l| l.sha3_256.clone()).collect();
    let sha256 = merkle_root_hex::<Sha256>(&sha256_leaves)?;
    let sha3_256 = merkle_root_hex::<Sha3_256>(&sha3_leaves)?;
    Some(DualHash { sha256, sha3_256 })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sha256_hex_known() {
        let hash = sha256_hex(b"hello");
        assert_eq!(hash, "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824");
    }

    #[test]
    fn test_sha3_256_hex_known() {
        let hash = sha3_256_hex(b"hello");
        assert_eq!(hash.len(), 64);
    }

    #[test]
    fn test_dual_hash_produces_both() {
        let dh = dual_hash(b"test");
        assert_eq!(dh.sha256.len(), 64);
        assert_eq!(dh.sha3_256.len(), 64);
        assert_ne!(dh.sha256, dh.sha3_256);
    }

    #[test]
    fn test_dual_merkle_root_empty() {
        assert!(dual_merkle_root(&[]).is_none());
    }

    #[test]
    fn test_dual_merkle_root_single_leaf() {
        let leaf = dual_hash(b"leaf-0");
        let root = dual_merkle_root(&[leaf.clone()]).unwrap();
        assert_eq!(root.sha256, leaf.sha256);
        assert_eq!(root.sha3_256, leaf.sha3_256);
    }

    #[test]
    fn test_dual_merkle_root_deterministic_and_non_trivial() {
        let leaves: Vec<DualHash> = (0..4).map(|i| dual_hash(format!("leaf-{i}").as_bytes())).collect();
        let r1 = dual_merkle_root(&leaves).unwrap();
        let r2 = dual_merkle_root(&leaves).unwrap();
        assert_eq!(r1, r2);
        assert_eq!(r1.sha256.len(), 64);
        assert_eq!(r1.sha3_256.len(), 64);
        assert_ne!(r1.sha256, r1.sha3_256);
    }
}
